hello-app-iap/app.py

```diff
@@ -0,0 +1,45 @@
+# Copyright 2017 Google Inc.
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#   http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+ 
+ 
+from http.server import HTTPServer, BaseHTTPRequestHandler
+import os
+PORT = int(os.environ.get("PORT", "8080"))
+class RequestHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
```

Comment:
This app will work just fine without IAP and that's not good. I recommend you do these 2:

1. detect the header, if missing return a simple text response: `error: application is not running behind Identity-Aware Proxy (x-dsfldsjkfdlsfjkas header missing)`

2. when you find the header use the methods described at htps://cloud.google.com/iap/docs/signed-headers-howto#verify_the_id_token_header to validate the claim. we know that for proxies coming from IAP, these headers CANNOT be forged. However, what if you're on a GKE cluster and another pod within the cluster queries your NodePort service directly by forging the headers?

Reply:
This is what ESP is for, it does this JWT validation against the claim for you.

Reply:
I think we should have multiple examples for IAP.
1. Pushing the responsibility of validating the JWT to the app.
2. Using ESP to validate the JWT (no user code change required)
3. Using Envoy to validate the JWT.

Reply:
But how does ESP help you prevent other pods in the cluster circumvent the external LB and call the internal pod directly? Thats why you need in-service JWT validation too, I think.

I’m a bit against talking about stuff like envoy in our tutorials just yet.